File: B2CReplacementDesigner.Server/Models/TrustFrameworkEntities.cs
namespace B2CReplacementDesigner.Server.Models
{
    /// <summary>
    /// Base class for all TrustFramework policy entities
    /// </summary>
    public abstract class TrustFrameworkEntity
    {
        /// <summary>
        /// Unique identifier of the entity
        /// </summary>
        public string Id { get; set; } = string.Empty;

        /// <summary>
        /// Type of the entity (ClaimType, TechnicalProfile, etc.)
        /// </summary>
        public string EntityType { get; set; } = string.Empty;

        /// <summary>
        /// Source file where this entity is defined
        /// </summary>
        public string SourceFile { get; set; } = string.Empty;

        /// <summary>
        /// Source policy ID where this entity is defined
        /// </summary>
        public string SourcePolicyId { get; set; } = string.Empty;

        /// <summary>
        /// XPath to this element in the source file
        /// </summary>
        public string XPath { get; set; } = string.Empty;

        /// <summary>
        /// Hierarchy depth in policy inheritance (0 = root)
        /// </summary>
        public int HierarchyDepth { get; set; }

        /// <summary>
        /// Whether this entity overrides one from a base policy
        /// </summary>
        public bool IsOverride { get; set; }

        /// <summary>
        /// Raw XML content of this entity
        /// </summary>
        public string RawXml { get; set; } = string.Empty;
    }

    /// <summary>
    /// Represents a ClaimType from the ClaimsSchema
    /// </summary>
    public class ClaimTypeEntity : TrustFrameworkEntity
    {
        public string? DisplayName { get; set; }
        public string? DataType { get; set; }
        public string? UserInputType { get; set; }
        public string? Mask { get; set; }
        public string? AdminHelpText { get; set; }
        public string? UserHelpText { get; set; }
        public Dictionary<string, string> DefaultPartnerClaimTypes { get; set; } = new();
        public RestrictionInfo? Restriction { get; set; }
        public string? PredicateValidationReference { get; set; }
    }

    public class RestrictionInfo
    {
        public string? Pattern { get; set; }
        public List<EnumerationItem> Enumeration { get; set; } = new();
    }

    public class EnumerationItem
    {
        public string Value { get; set; } = string.Empty;
        public string Text { get; set; } = string.Empty;
        public bool SelectByDefault { get; set; }
    }

    /// <summary>
    /// Represents a TechnicalProfile
    /// </summary>
    public class TechnicalProfileEntity : TrustFrameworkEntity
    {
        public string? DisplayName { get; set; }
        public string? Description { get; set; }
        public string? ProtocolName { get; set; }
        public string? ProtocolHandler { get; set; }
        public string? ProviderName { get; set; }
        public string? ClaimsProviderDisplayName { get; set; }
        public List<MetadataItemInfo> Metadata { get; set; } = new();
        public List<string> InputClaimsTransformations { get; set; } = new();
        public List<ClaimReferenceInfo> InputClaims { get; set; } = new();
        public List<ClaimReferenceInfo> DisplayClaims { get; set; } = new();
        public List<string> DisplayControls { get; set; } = new();
        public List<ClaimReferenceInfo> PersistedClaims { get; set; } = new();
        public List<ClaimReferenceInfo> OutputClaims { get; set; } = new();
        public List<string> OutputClaimsTransformations { get; set; } = new();
        public List<string> ValidationTechnicalProfiles { get; set; } = new();
        public string? IncludeTechnicalProfile { get; set; }
        public List<InheritanceInfo> InheritanceChain { get; set; } = new();
    }

    public class MetadataItemInfo
    {
        public string Key { get; set; } = string.Empty;
        public string Value { get; set; } = string.Empty;
    }

    public class ClaimReferenceInfo
    {
        public string? ClaimTypeReferenceId { get; set; }
        public string? PartnerClaimType { get; set; }
        public string? DefaultValue { get; set; }
        public bool? AlwaysUseDefaultValue { get; set; }
        public bool? Required { get; set; }
        public string? DisplayControlReferenceId { get; set; }
    }

    public class InheritanceInfo
    {
        public string ProfileId { get; set; } = string.Empty;
        public string PolicyId { get; set; } = string.Empty;
        public string FileName { get; set; } = string.Empty;
        public InheritanceType InheritanceType { get; set; }
    }

    public enum InheritanceType
    {
        Direct,
        Include
    }

    /// <summary>
    /// Represents a ClaimsTransformation
    /// </summary>
    public class ClaimsTransformationEntity : TrustFrameworkEntity
    {
        public string TransformationMethod { get; set; } = string.Empty;
        public List<ClaimReferenceInfo> InputClaims { get; set; } = new();
        public List<InputParameterInfo> InputParameters { get; set; } = new();
        public List<ClaimReferenceInfo> OutputClaims { get; set; } = new();
    }

    public class InputParameterInfo
    {
        public string Id { get; set; } = string.Empty;
        public string DataType { get; set; } = string.Empty;
        public string Value { get; set; } = string.Empty;
    }

    /// <summary>
    /// Represents a DisplayControl
    /// </summary>
    public class DisplayControlEntity : TrustFrameworkEntity
    {
        public string UserInterfaceControlType { get; set; } = string.Empty;
        public List<ClaimReferenceInfo> InputClaims { get; set; } = new();
        public List<ClaimReferenceInfo> DisplayClaims { get; set; } = new();
        public List<ClaimReferenceInfo> OutputClaims { get; set; } = new();
        public List<DisplayControlActionInfo> Actions { get; set; } = new();
    }

    public class DisplayControlActionInfo
    {
        public string Id { get; set; } = string.Empty;
        public List<string> ValidationClaimsExchanges { get; set; } = new();
    }

    /// <summary>
    /// Represents a UserJourney
    /// </summary>
    public class UserJourneyEntity : TrustFrameworkEntity
    {
        public string? DefaultCpimIssuerTechnicalProfileReferenceId { get; set; }
        public List<OrchestrationStepInfo> OrchestrationSteps { get; set; } = new();
    }

    public class OrchestrationStepInfo
    {
        public int Order { get; set; }
        public string Type { get; set; } = string.Empty;
        public string? ContentDefinitionReferenceId { get; set; }
        public List<ClaimsExchangeInfo> ClaimsExchanges { get; set; } = new();
        public List<PreconditionInfo> Preconditions { get; set; } = new();
    }

    public class ClaimsExchangeInfo
    {
        public string Id { get; set; } = string.Empty;
        public string TechnicalProfileReferenceId { get; set; } = string.Empty;
    }

    public class PreconditionInfo
    {
        public string Type { get; set; } = string.Empty;
        public bool ExecuteActionsIf { get; set; }
        public List<string> Values { get; set; } = new();
        public string Action { get; set; } = string.Empty;
    }

    /// <summary>
    /// Represents a SubJourney
    /// </summary>
    public class SubJourneyEntity : TrustFrameworkEntity
    {
        public string Type { get; set; } = string.Empty;
        public List<OrchestrationStepInfo> OrchestrationSteps { get; set; } = new();
    }

    /// <summary>
    /// Represents a ClaimsProvider
    /// </summary>
    public class ClaimsProviderEntity : TrustFrameworkEntity
    {
        public string? DisplayName { get; set; }
        public List<string> TechnicalProfileIds { get; set; } = new();
    }

    /// <summary>
    /// Collection of all extracted policy entities
    /// </summary>
    public class PolicyEntities
    {
        public Dictionary<string, List<ClaimTypeEntity>> ClaimTypes { get; set; } = new();
        public Dictionary<string, List<TechnicalProfileEntity>> TechnicalProfiles { get; set; } = new();
        public Dictionary<string, List<ClaimsTransformationEntity>> ClaimsTransformations { get; set; } = new();
        public Dictionary<string, List<DisplayControlEntity>> DisplayControls { get; set; } = new();
        public Dictionary<string, List<UserJourneyEntity>> UserJourneys { get; set; } = new();
        public Dictionary<string, List<SubJourneyEntity>> SubJourneys { get; set; } = new();
        public Dictionary<string, List<ClaimsProviderEntity>> ClaimsProviders { get; set; } = new();
    }
}

File: client/src/types/trust-framework-entities.ts
import type { ProtocolName } from '@/types/technical-profile';

/**
 * Comprehensive type definitions for all Azure AD B2C TrustFramework entities
 * Based on TrustFrameworkPolicy_0.3.0.0.xsd schema and backend models
 * 
 * ## Important: Entity Storage Structure
 * 
 * All entities are stored as arrays to preserve the inheritance chain.
 * Each entity ID maps to an array of entity versions from different policy files.
 * 
 * Example:
 * ```typescript
 * entities.technicalProfiles['SelfAsserted-LocalAccountSignin-Email'] = [
 *   { ...entity from Base.xml, hierarchyDepth: 0 },
 *   { ...entity from Extension.xml, hierarchyDepth: 1 },
 *   { ...entity from SignUpOrSignIn.xml, hierarchyDepth: 2 }
 * ]
 * ```
 * 
 * ## Working with Entities
 * 
 * ### Default Display (Consolidated View)
 * Use `getEntity()` to get the "default" entity for UI display:
 * - Returns the entity from ConsolidatedPolicy if available
 * - Otherwise returns the first entity in the array
 * 
 * ```typescript
 * const profile = getEntity(entities, 'TechnicalProfile', 'SelfAsserted-LocalAccountSignin-Email');
 * // This is what you display in the graph, nodes, and default sidebar
 * ```
 * 
 * ### Inheritance Chain Navigation
 * Use `getAllEntities()` to get all versions for inheritance navigation:
 * 
 * ```typescript
 * const allVersions = getAllEntities(entities, 'TechnicalProfile', 'SelfAsserted-LocalAccountSignin-Email');
 * // Use this when user clicks through the inheritance chain to see different policy versions
 * allVersions.forEach(version => {
 *   console.log(`${version.sourcePolicyId}: ${version.sourceFile}`);
 * });
 * ```
 * 
 * ### Best Practices
 * - **Always use helper functions** (`getEntity`, `getAllEntities`, `hasEntity`) instead of direct dictionary access
 * - **Default display**: Use `getEntity()` for graph rendering and initial sidebar display
 * - **Inheritance navigation**: Use `getAllEntities()` when user navigates through policy hierarchy
 * - **ConsolidatedPolicy** is the "merged" view and should be the default when available
 */

/**
 * Base interface for all TrustFramework policy entities
 */
export interface TrustFrameworkEntity {
    id: string;
    entityType: string;
    sourceFile: string;
    sourcePolicyId: string;
    xpath: string;
    hierarchyDepth: number;
    isOverride: boolean;
    rawXml: string;
}

/**
 * Enumeration item for claim restrictions
 */
export interface EnumerationItem {
    value: string;
    text: string;
    selectByDefault: boolean;
}

/**
 * Restriction information for claims
 */
export interface RestrictionInfo {
    pattern?: string;
    enumeration: EnumerationItem[];
}

/**
 * ClaimType entity from ClaimsSchema
 */
export interface ClaimTypeEntity extends TrustFrameworkEntity {
    entityType: 'ClaimType';
    displayName?: string;
    dataType?: string;
    userInputType?: string;
    mask?: string;
    adminHelpText?: string;
    userHelpText?: string;
    defaultPartnerClaimTypes: Record<string, string>;
    restriction?: RestrictionInfo;
    predicateValidationReference?: string;
}

/**
 * Metadata item
 */
export interface MetadataItemInfo {
    key: string;
    value: string;
}

/**
 * Claim reference information
 */
export interface ClaimReferenceInfo {
    claimTypeReferenceId?: string;
    partnerClaimType?: string;
    defaultValue?: string;
    alwaysUseDefaultValue?: boolean;
    required?: boolean;
    displayControlReferenceId?: string;
}

/**
 * Inheritance information
 */
export interface InheritanceInfo {
    profileId: string;
    policyId: string;
    fileName: string;
    inheritanceType: 'Direct' | 'Include';
}

/**
 * TechnicalProfile entity
 */
export interface TechnicalProfileEntity extends TrustFrameworkEntity {
    entityType: 'TechnicalProfile';
    displayName?: string;
    description?: string;
    protocolName?: ProtocolName;
    protocolHandler?: string;
    claimsProviderDisplayName?: string;
    metadata: MetadataItemInfo[];
    inputClaimsTransformations: string[];
    inputClaims: ClaimReferenceInfo[];
    displayClaims: ClaimReferenceInfo[];
    displayControls: string[];
    persistedClaims: ClaimReferenceInfo[];
    outputClaims: ClaimReferenceInfo[];
    outputClaimsTransformations: string[];
    validationTechnicalProfiles: string[];
    includeTechnicalProfile?: string;
    inheritanceChain: InheritanceInfo[];
}

/**
 * Input parameter for claims transformation
 */
export interface InputParameterInfo {
    id: string;
    dataType: string;
    value: string;
}

/**
 * ClaimsTransformation entity
 */
export interface ClaimsTransformationEntity extends TrustFrameworkEntity {
    entityType: 'ClaimsTransformation';
    transformationMethod: string;
    inputClaims: ClaimReferenceInfo[];
    inputParameters: InputParameterInfo[];
    outputClaims: ClaimReferenceInfo[];
}

/**
 * Display control action information
 */
export interface DisplayControlActionInfo {
    id: string;
    validationClaimsExchanges: string[];
}

/**
 * DisplayControl entity
 */
export interface DisplayControlEntity extends TrustFrameworkEntity {
    entityType: 'DisplayControl';
    userInterfaceControlType: string;
    inputClaims: ClaimReferenceInfo[];
    displayClaims: ClaimReferenceInfo[];
    outputClaims: ClaimReferenceInfo[];
    actions: DisplayControlActionInfo[];
}

/**
 * Claims exchange information
 */
export interface ClaimsExchangeInfo {
    id: string;
    technicalProfileReferenceId: string;
}

/**
 * Precondition information
 */
export interface PreconditionInfo {
    type: string;
    executeActionsIf: boolean;
    values: string[];
    action: string;
}

/**
 * Orchestration step information
 */
export interface OrchestrationStepInfo {
    order: number;
    type: string;
    contentDefinitionReferenceId?: string;
    claimsExchanges: ClaimsExchangeInfo[];
    preconditions: PreconditionInfo[];
}

/**
 * UserJourney entity
 */
export interface UserJourneyEntity extends TrustFrameworkEntity {
    entityType: 'UserJourney';
    defaultCpimIssuerTechnicalProfileReferenceId?: string;
    orchestrationSteps: OrchestrationStepInfo[];
}

/**
 * SubJourney entity
 */
export interface SubJourneyEntity extends TrustFrameworkEntity {
    entityType: 'SubJourney';
    type: string;
    orchestrationSteps: OrchestrationStepInfo[];
}

/**
 * ClaimsProvider entity
 */
export interface ClaimsProviderEntity extends TrustFrameworkEntity {
    entityType: 'ClaimsProvider';
    displayName?: string;
    technicalProfileIds: string[];
}

/**
 * Collection of all policy entities
 */
export interface PolicyEntities {
    claimTypes: Record<string, ClaimTypeEntity[]>;
    technicalProfiles: Record<string, TechnicalProfileEntity[]>;
    claimsTransformations: Record<string, ClaimsTransformationEntity[]>;
    displayControls: Record<string, DisplayControlEntity[]>;
    userJourneys: Record<string, UserJourneyEntity[]>;
    subJourneys: Record<string, SubJourneyEntity[]>;
    claimsProviders: Record<string, ClaimsProviderEntity[]>;
}

/**
 * Union type of all entity types
 */
export type AnyTrustFrameworkEntity =
    | ClaimTypeEntity
    | TechnicalProfileEntity
    | ClaimsTransformationEntity
    | DisplayControlEntity
    | UserJourneyEntity
    | SubJourneyEntity
    | ClaimsProviderEntity;

/**
 * Entity type discriminator
 */
export type EntityType =
    | 'ClaimType'
    | 'TechnicalProfile'
    | 'ClaimsTransformation'
    | 'DisplayControl'
    | 'UserJourney'
    | 'SubJourney'
    | 'ClaimsProvider';

/**
 * Constants for consolidated policy identification
 */
export const CONSOLIDATED_FILE_NAME = 'ConsolidatedPolicy';
export const CONSOLIDATED_POLICY_ID = 'Consolidated';

/**
 * Helper to get the default (consolidated) entity from an array of entities.
 * Returns the entity from the consolidated policy if available, otherwise the first entity.
 */
function getDefaultEntity<T extends TrustFrameworkEntity>(
    entities: T[] | undefined
): T | undefined {
    if (!entities || entities.length === 0) return undefined;
    
    const consolidated = entities.find(
        e => e.sourceFile === CONSOLIDATED_FILE_NAME || 
             e.sourcePolicyId === CONSOLIDATED_POLICY_ID
    );
    
    return consolidated || entities[0];
}

/**
 * Helper to get all entities from entities collection by type and ID
 */
export function getAllEntities(
    entities: PolicyEntities,
    type: EntityType,
    id: string
): AnyTrustFrameworkEntity[] {
    switch (type) {
        case 'ClaimType':
            return entities.claimTypes[id] || [];
        case 'TechnicalProfile':
            return entities.technicalProfiles[id] || [];
        case 'ClaimsTransformation':
            return entities.claimsTransformations[id] || [];
        case 'DisplayControl':
            return entities.displayControls[id] || [];
        case 'UserJourney':
            return entities.userJourneys[id] || [];
        case 'SubJourney':
            return entities.subJourneys[id] || [];
        case 'ClaimsProvider':
            return entities.claimsProviders[id] || [];
        default:
            return [];
    }
}

/**
 * Helper to get the default entity from entities collection by type and ID.
 * Returns the consolidated entity if available, otherwise the first entity.
 */
export function getEntity(
    entities: PolicyEntities,
    type: EntityType,
    id: string
): AnyTrustFrameworkEntity | undefined {
    const allEntities = getAllEntities(entities, type, id);
    return getDefaultEntity(allEntities);
}

/**
 * Helper to check if an entity exists
 */
export function hasEntity(
    entities: PolicyEntities,
    type: EntityType,
    id: string
): boolean {
    return getEntity(entities, type, id) !== undefined;
}
